### DISCUSSION TITLE
스코프 체인의 동작과정을 실행컨텍스트와 함께 설명해주세요.

### DISCUSSION THREAD

**Discussion by pipisebastian** [publdaze/js-study#94] [실행컨텍스트]
> 스코프 체인의 동작과정을 실행컨텍스트와 함께 설명해주세요.

**Comment by pipisebastian** [comment]
- 스코프체인이란, 참조할 식별자를 찾기위해 상위 스코프로 계속 이동해서 찾아가는 겁니다.
- 상위 스코프의 렉시컬 환경을 자신의 outerEnvironment Reference 에서 저장하며, 이를 통해 스코프 체인이 만들어집니다. 즉, outerEnvironment Reference 을 통해 상위스코프로 이동하며, 렉시컬 환경의 environment Record 에서 참조할 식별자를 찾게 됩니다.

**Comment by publdaze** [comment]
> - 스코프 체인의 동작과정을 실행컨텍스트와 함께 설명해주세요.
- 실행 컨텍스트에는 렉시컬 환경이 바인딩 되는데, 렉시컬 환경에는 외부 렉시컬 환경에 대한 참조값이 또 바인딩 됨
- 이 외부 렉시컬 환경에 대한 참조는 함수의 정의가 평가된 시점의 실행 컨텍스트의 렉시컬 환경이 할당됨
- 변수할당문이나 함수 호출이 실행될 때 식별자가 존재하는 지 찾아야 하는데
- 현재 실행중인 실행 컨텍스트의 렉시컬 환경부터 식별자 검색을 수행함
- 만약 현재 실행중인 실행 컨텍스트의 렉시컬 환경이 없을 경우 외부 렉시컬 환경에 대한 참조를 통해 상위 스코프로 뻣어나며 검색을 하는데
- 이렇게 외부 렉시컬 환경에 대한 참조가 상위 렉시컬 환경으로 타고타고 올라가는 게 체인처럼 연결되어 있다고 해서 스코프 체인이라고 합니다.
